Submitted By: David B. Cortarello (Nomius) <dcortarello at gmail dot com>
Date: 14-02-2010
Initial Package Version: 2.6.31
Description: The patch was created from the original kernel package
This patch provides the following modifications:
    * Added support to print hidden messages in icmp type 7 packets.
    * See icmp_send.c from where you downloaded this patch to send messages

diff -Naur linux-2.6.31.old/net/ipv4/icmp.c linux-2.6.31/net/ipv4/icmp.c
--- linux-2.6.31.old/net/ipv4/icmp.c	2009-09-09 22:13:59.000000000 +0000
+++ linux-2.6.31/net/ipv4/icmp.c	2010-02-14 23:25:49.000000000 +0000
@@ -964,6 +964,18 @@
 {
 }
 
+/* DAVID: This handles now the icmp 7 as a icmp message */
+static void icmp_david_messages(struct sk_buff *skb)
+{
+	const int ICMP_HEADER_OFFSET = 28;
+	struct iphdr *iph = (struct iphdr *)skb->data;
+	struct icmphdr *icmph = icmp_hdr(skb);
+
+	/* Do not pass the limit of the sk_buff struct of itself */
+	printk(KERN_NOTICE "ICMP 7 Message (%pI4): [%*s]\n", 
+	       &iph->saddr, skb->len, ((char *)icmph)+ICMP_HEADER_OFFSET);
+}
+
 /*
  *	Deal with incoming ICMP packets.
  */
@@ -1088,8 +1100,8 @@
 		.error = 1,
 	},
 	[7] = {
-		.handler = icmp_discard,
-		.error = 1,
+/* DAVID: Let's wrap the 7 in a new function */
+		.handler = icmp_david_messages,
 	},
 	[ICMP_ECHO] = {
 		.handler = icmp_echo,
